### PR TITLE
STORM-3822 fix visualization when streamId contains colon

### PR DESCRIPTION
**Pull request by ben-roling**
## What is the purpose of the change

This fixes [STORM-3822](https://issues.apache.org/jira/browse/STORM-3822), making the Topology Visualization UI render correctly when a streamId contains a colon character.

## How was the change tested

I have tested with a modified version of WordCountTopology where I used a custom streamId ("my:stream") containing a colon and verified the UI loads successfully.  I have also tested with some of my team's real topologies that utilize streamIds containing colons.

Further, I have added a unit test for the function of the sanitizeStreamName() method modified in this PR.

## Notes

The sanitizeStreamName() method body was a bit cryptic and hard to follow.  I refactored it in an attempt to make it more readable.  I could have gone a little further.  In the initial version in the PR I have just added comments, renamed variables, and created an if/else block where previously there was a default assignment followed by a reassignment in an if condition.

The real change I made was simply to remove ':' as one of the allowed (not replaced) characters such that it would be replaced by '_' like other classes of characters that aren't allowed.

I have also done a couple of other things discussed in [STORM-3817](https://issues.apache.org/jira/browse/STORM-3817):
* removed yetus dependency exclusion that was causing me local build failures
* cherry-picked 1988a47f2b82ec103fffb02fc323aab0de3b09c6 so hopefully travis will build successfully

These last two items can be removed from this PR before merge, assuming they get incorporated elsewhere.

### CODE REVIEW

File: storm-webapp/src/main/java/org/apache/storm/daemon/ui/UIHelpers.java [hunk 1]
```diff
@@ -1746,19 +1746,26 @@ private static Double nullToZero(Double value) {
...
      */
     public static String sanitizeStreamName(String streamName) {
-        Pattern pattern = Pattern.compile("(?![A-Za-z_\\-:\\.]).");
-        Pattern pattern2 = Pattern.compile("^[A-Za-z]");
-        Matcher matcher = pattern2.matcher(streamName);
-        Matcher matcher2 = pattern.matcher("\\s" + streamName);
-        if (matcher.find()) {
-            matcher2 = pattern.matcher(streamName);
+        Pattern startsWithLetterPattern = Pattern.compile("^[A-Za-z]");
+        Pattern problemCharacterPattern = Pattern.compile("(?![A-Za-z_\\-\\.]).");
```
[the remark this answers is not in the document]
Comment (ben-roling) [reply storm-webapp/src/main/java/org/apache/storm/daemon/ui/UIHelpers.java:1756]: The only "real" change I made was to remove the colon character from this pattern (previously named `pattern`).

**Review by rzo1 — APPROVED** [review]

File: .travis.yml [hunk 2]
```diff
@@ -25,7 +25,7 @@ env:
   - MODULES=Integration-Test
   - MODULES=Check-Updated-License-Files
 
-dist: xenial
+dist: bionic
```
Comment (agresch) [inline .travis.yml:28]: This should be put up as a separate JIRA if needed.

File: storm-shaded-deps/pom.xml [hunk 3]
```diff
@@ -71,24 +71,12 @@
                     <groupId>jline</groupId>
                     <artifactId>jline</artifactId>
                 </exclusion>
```
Comment (agresch) [inline storm-shaded-deps/pom.xml:73]: This should be put up as a separate JIRA if needed.

**Comment by rzo1** [comment]
@ben-roling think this needs a rebase with master.

**Comment by ben-roling** [comment]
@agresch @rzo1, I've rebased leaving in only the essential changes.  We'll see how things go with the travis build.

File: storm-webapp/src/test/java/org/apache/storm/daemon/ui/UIHelpersTest.java [hunk 4]
```diff
@@ -485,6 +485,19 @@ void test_getTopologySpoutAggStatsMap_generalFields() {
         assertEquals("", spoutResult.get("lastError"), "No error should be reported as empty string");
     }
 
+    /**
+     * Tests that santizeStreamName() does the expected manipulations
+     */
+    @Test
+    public void testSanitizeStreamName() {
```
Comment (bipinprasad) [inline storm-webapp/src/test/java/org/apache/storm/daemon/ui/UIHelpersTest.java:492]: Thanks for adding the test case

File: storm-webapp/src/main/java/org/apache/storm/daemon/ui/UIHelpers.java [hunk 5]
```diff
@@ -1746,19 +1746,26 @@ private static Double nullToZero(Double value) {
...
+
+        Matcher startsWithLetterMatcher = startsWithLetterPattern.matcher(streamName);
+        Matcher problemCharacterMatcher;
+        if (startsWithLetterMatcher.find()) {
+            problemCharacterMatcher = problemCharacterPattern.matcher(streamName);
+        } else {
+            // if the streamName starts with non-letter character, prepend "\s".  The backslash will then
+            // get replaced by an underscore, resulting in the prefix of "_s" in the final sanitized name
+            problemCharacterMatcher = problemCharacterPattern.matcher("\\s" + streamName);
```
[the remark this answers is not in the document]
Comment (ben-roling) [reply storm-webapp/src/main/java/org/apache/storm/daemon/ui/UIHelpers.java]: I've added the simplification as you suggest.  Perhaps I should have done that straight away.

https://github.com/apache/storm/pull/3441/commits/b7f04e7e6d9d0b05d449795279fd8c250cd92abf

File: storm-webapp/src/main/java/org/apache/storm/daemon/ui/UIHelpers.java [hunk 6]
```diff
@@ -1746,19 +1746,21 @@ private static Double nullToZero(Double value) {
...
     public static String sanitizeStreamName(String streamName) {
-        Pattern pattern = Pattern.compile("(?![A-Za-z_\\-:\\.]).");
-        Pattern pattern2 = Pattern.compile("^[A-Za-z]");
-        Matcher matcher = pattern2.matcher(streamName);
-        Matcher matcher2 = pattern.matcher("\\s" + streamName);
-        if (matcher.find()) {
-            matcher2 = pattern.matcher(streamName);
-        }
-        return matcher2.replaceAll("_");
+        if (streamName == null || streamName.isEmpty()) {
```
Comment (ben-roling) [inline storm-webapp/src/main/java/org/apache/storm/daemon/ui/UIHelpers.java:1755]: It doesn't seem the streamName would ever be null or empty but I added this precondition checking to fail faster if it ever was.  Without it the exception could occur at `Character.isLetter(streamName.charAt(0)`.  I can delete this and let that happen instead if there is a preference.

File: storm-webapp/src/main/java/org/apache/storm/daemon/ui/UIHelpers.java [hunk 7]
```diff
@@ -1746,19 +1746,21 @@ private static Double nullToZero(Double value) {
...
+            throw new IllegalArgumentException("streamName: " + streamName == null ? "null" : "empty");
```
Comment (bipinprasad) [inline storm-webapp/src/main/java/org/apache/storm/daemon/ui/UIHelpers.java:1756]: Since this is used in the UI, it is better to return the string as is when null or empty.

Comment (ben-roling) [reply storm-webapp/src/main/java/org/apache/storm/daemon/ui/UIHelpers.java]: I've restored parity with how null and empty were handled before any of my changes:
https://github.com/apache/storm/pull/3441/commits/177976b929bf18dd0542bb4131cd801e99a827f7

If streamName is null, it will result in NPE just as it did before.  I don't think there is any way streamName would ever be null.  I don't think it would ever be empty either, but since I have at least a small degree of uncertainty this way we can be sure it at least behaves the same way it used to.

**Comment by bipinprasad** [comment]
I restarted the failed check.
